Refactor builder with transactions

s9latimm opened the pull request:
Implement FMBuilder using transactions while adding required modifications.

- [x] Refactor `FeatureModelBuilder`
  - [x] move to own file
  - [x] `AddChild`
  - [x] `SetRoot`
  - [x] `SetPath`
  - [x] `SetCommit`
  - [x] `SetName`
  - [x] `AddConstraintToModel` (see se-passau/VaRA/issues/664)
  - [x] `AddRelationshipToModel` (see #44)
    - [x] tests 
- [x] Fix failing tests
- [x] Clang-Tidy

Related se-passau/VaRA/issues/744

Closes se-passau/VaRA/issues/635
Closes se-passau/VaRA/issues/669
Closes se-passau/VaRA/issues/692
Closes se-passau/VaRA/issues/724

s9latimm committed:
WIP: Refactor builder

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
A	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/CMakeLists.txt
M	lib/Feature/FeatureModel.cpp
A	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Feature/OrderedFeatureVector.cpp
M	unittests/Feature/Relationship.cpp
M	unittests/Feature/RootFeature.cpp

s9latimm committed:
WIP: Add edge mods

M	include/vara/Feature/FeatureModelTransaction.h

Commit:
WIP feature model merge

Commit:
tidy fixes

Commit:
tidy fixes for llvm11

s9latimm committed:
WIP: Failing tests

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/RootFeature.cpp

s9latimm committed:
WIP: Use transactions

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

Commit:
WIP Transactions can add relationships

s9latimm committed:
WIP: Support Relationships (see #44)

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModel.cpp

s9latimm committed:
Add constraints and most stuff working

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Feature/RootFeature.cpp

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-FMBuilderMods

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/feature/test_location.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	bindings/python/vara-feature/pybind_FeatureSourceRange.cpp
M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModelParser.cpp

Commit:
Implement feedback

Commit:
Implement feedback

Remove use of reference in addLocation

Commit:
Merge remote-tracking branch 'origin/f-FMBuilderMods' into f-fmMerge

# Conflicts:
#	include/vara/Feature/FeatureModel.h
#	include/vara/Feature/FeatureModelTransaction.h

Commit:
Fixes

padupr committed:
Update lib/Feature/FeatureModelTransaction.cpp

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

Commit:
Allow to add relationships without specifying children

Commit:
Merge remote-tracking branch 'origin/f-fmMerge' into f-fmMerge

vulder committed:
Adds delegating constructor for FeatureSourceRanges

Allow FSR to be constructed from temporaries.

s9latimm committed:
Fix relationships

M	.gitignore
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/Relationship.cpp

s9latimm committed:
Fix tests

M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Fix tidy

M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/FeatureModelTransaction.cpp

vulder commented on include/vara/Feature/Constraint.h:
```diff
@@ -595,7 +599,7 @@ class PrimaryFeatureConstraint : public PrimaryConstraint {
 
 private:
-  friend FeatureModelBuilder;
+  friend detail::FeatureModelModification;
```
please move to top of the class

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -274,4 +275,16 @@ class RootFeature : public Feature {
 } // namespace vara::feature
 
+inline std::ostream &operator<<(std::ostream &Out,
+                                const vara::feature::Feature *Feature) {
```
```suggestion
                                const vara::feature::Feature &Feature) {
```
Should take the feature via `&` because we cannot handle a null state in the function and don't need to reassign the value.

vulder commented on include/vara/Feature/Feature.h:
```diff
@@ -274,4 +275,16 @@ class RootFeature : public Feature {
+}
+
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &Out,
+                                     const vara::feature::Feature *Feature) {
```
```suggestion
                                     const vara::feature::Feature &Feature) {
```

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -34,12 +34,12 @@ class FeatureModel {
-  using RelationshipContainerTy = std::vector<std::unique_ptr<RelationshipTy>>;
+  using RelationshipContainerTy = std::vector<std::unique_ptr<Relationship>>;
 
+  FeatureModel() = default;
```
Why do we need a public default ctor for FM?
The default constructed FM would not make sense because key values are empty.

s9latimm replied:
This is kind of required, if we want to build a FM only using transactions, so we start with an empty tree.
It may be reasonable to initialize a root feature at this point, which would make that explicit step in the builder obsolete.
Currently completely empty FM (without any nodes and especially without a root) are valid.

s9latimm replied:
I think making the FMbuilder a friend and this private may be a better approach.

We might want to support a constructor without features, constraints etc though.

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -58,9 +58,10 @@ class FeatureModel {
+  [[nodiscard]] RootFeature *getRoot() const { return Root; }
 
-  [[nodiscard]] Feature *getRoot() const { return Root; }
+  // TODO make this private and use a transaction
```
TODO

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -130,149 +129,31 @@ class FeatureModel {
-  /// represented als mutual excluded but non-optional features (which requires
-  /// additional processing).
-  void detectXMLAlternatives();
+  void sort() { OrderedFeatures.sort(); }
```
Is this ever needed? Shouldn't the OrderedFeatureVector guarantee to be always ordered?

s9latimm replied:
The ordering inside the vector is currently based on the hierarchy of a feature at the moment it is inserted.
So this is currently not dynamically updated.
As we can change the structure of the FM through transaction, we need to reevaluate this sometimes (for example based on changed parents).

I think we should provide this constraint with the collection itself -- will look into it.

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -19,10 +20,14 @@ class FeatureModelCopyTransactionBase;
                                    CheckFeatureParentChildRelationShip>;
 } // namespace detail
 
+using FeatureVariant = std::variant<std::string, Feature *>;
+using FeatureTreeNodeVariant = std::variant<std::string, FeatureTreeNode *>;
```
What is the idea behind this 2 global var defs?

s9latimm replied:
I found it suitable to provide transactions with the ability to handle features both based on name or pointer.
In most cases the provided feature may not exist yet and only be created while committing, so we can only refer to it by name -- not sure if supporting pointers has huge benefits.

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -58,7 +63,12 @@ class FeatureModelTransaction
+                                  // changes are committed before destruction
+        llvm::errs()
+            << "warning: Uncommitted modifications before destruction.\n";
+        // TODO(s9latimm): Committing now may break with prior failed commits.
```
Set in on the agenda for friday

vulder replied:
@s9latimm What do we do here?

s9latimm replied:
This strongly depends on se-passau/VaRA/issues/744, as we need a way to track if a prior commit failed to prevent breaking the FM further and prohibit unintended behavior -- this may be combined with a rollback (see se-passau/VaRA/issues/723) while discarding any uncommitted changes or even require success when invoked without a prior commit.

I kept this open for now and linked the related issue.

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -163,6 +255,14 @@ class FeatureModelModification {
     FM.removeFeature(F);
   }
 
+  template <class PtrTy>
+  static PtrTy *resolveVariant(FeatureModel &FM,
+                               std::variant<std::string, PtrTy *> V) {
+    return std::holds_alternative<std::string>(V)
+               ? FM.getFeature(std::get<std::string>(V))
+               : std::get<PtrTy *>(V);
+  }
```
The name is a bit confusing and unspecific.
I also think, this method is not necessary. If anyone needs to handle a variant he should use `visit` to handle the different types.

s9latimm replied:
This may be obsolete, if we decide that transactions only take names of features and not pointers (see above).

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+      return nullptr;
+    }
+
+    auto *P = resolveVariant(FM, Parent);
```
This should use visit

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+public:
+  void exec(FeatureModel &FM) override { (*this)(FM); }
+
+  void operator()(FeatureModel &FM) { setName(FM, Name); }
```
You could move in the name

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+public:
+  void exec(FeatureModel &FM) override { (*this)(FM); }
+
+  void operator()(FeatureModel &FM) { setCommit(FM, Commit); }
```
same with commit

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+public:
+  void exec(FeatureModel &FM) override { (*this)(FM); }
+
+  void operator()(FeatureModel &FM) { setPath(FM, Path); }
```
same with path

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+      }
+      setRoot(FM, *NewRoot);
+    }
+    sort(FM);
```
Why is an explicit sort call needed here?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+
+  void operator()(FeatureModel &FM) {
+    auto *C = resolveVariant(FM, Child);
+    auto *P = resolveVariant(FM, Parent);
```
Visit

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -205,15 +308,235 @@ class AddFeatureToModel : public FeatureModelModification {
+    removeEdge(*C->getParent(), *C);
+    addEdge(*P, *C);
+    setParent(*C, *P);
+    sort(FM);
```
?

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -232,35 +555,101 @@ class FeatureModelCopyTransactionBase {
+
+  [[nodiscard]] Feature *TranslateFeature(Feature &F) {
+    return FM->getFeature(F.getName());
+  }
```
```suggestion
  [[nodiscard]] Feature *TranslateFeature(Feature &F) {
    return FM->getFeature(F.getName());
  }
  
  std::unique_ptr<FeatureModel> FM;
```
Member vars should be last

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -278,13 +667,97 @@ class FeatureModelModifyTransactionBase {
+  }
+
+  void setNameImpl(std::string Name) {
+    assert(FM && "");
```
We should fill the asserts with an useful message

vulder commented on include/vara/Feature/OrderedFeatureVector.h:
```diff
@@ -57,6 +59,13 @@ class OrderedFeatureVector {
     insert(Init.begin(), Init.end());
   }
 
+  void sort() {
```
same as mentioned above

vulder commented on lib/Feature/FeatureModelBuilder.cpp:
```diff
@@ -0,0 +1,89 @@
+
+/// Decide whether two features are mutual exclusive. Beware that this method
+/// only detects very simple trees with binary excludes.
+bool isSimpleMutex(const Feature *A, const Feature *B) {
```
I think, we should simplify this implementation a bit by refactoring out some parts. This is quite hard to understand.

vulder commented on lib/Feature/FeatureModelBuilder.cpp:
```diff
@@ -0,0 +1,89 @@
+    auto Children = F->getChildren<Feature>();
+    if (Children.size() > 1 &&
+        std::all_of(Children.begin(), Children.end(), [Children](auto *F) {
+          for (auto *C : Children) {
+            if (F == C) {
+              continue;
+            }
+            if (!isSimpleMutex(F, C)) {
+              return false;
+            }
+          }
+          return !F->isOptional();
```
This could probability be simplified by using a second `all_of`

vulder commented on lib/Feature/FeatureModelBuilder.cpp:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+  return Features.commit() && Transactions.commit() &&
+                 PostTransactions.commit() && detectXMLAlternatives() &&
```
What are PostTransactions?

vulder commented on lib/Feature/FeatureModelBuilder.cpp:
```diff
@@ -0,0 +1,89 @@
+      return nullptr;
+    }
+  }
+  return Features.commit() && Transactions.commit() &&
```
What is the meaning of Transactions here?

s9latimm replied:
These are dependent stages, if we want to assure, all feature reference can be resolved.

1. `Features` are all emplacements of feature nodes into the FM, without any structural information yet -- basically everything is child of root.
2. `Transactions` contains all modification to build the tree structure of features, like parent / child relations.
3. `PostTransactions` (we may use a better name, see above) insert relationships an constraints. At this stage we can assume, that all feature nodes and edges are already present, so we can easily change the hierarchy or let constraints refer to existing features (through this could also be done in the prior stage).
4. `Special` currently only handles XML alternative groups, so we could move this locally, but I think natively providing a collection for similar corner cases might be useful.

vulder commented on lib/Feature/FeatureModelBuilder.cpp:
```diff
@@ -0,0 +1,89 @@
+      return nullptr;
+    }
+  }
+  return Features.commit() && Transactions.commit() &&
```
Hmm, I think we could just chain the commits sequentially instead of pressing everything into one big && expression

vulder commented on unittests/Feature/FeatureModelBuilder.cpp:
```diff
@@ -126,17 +126,19 @@ TEST(FeatureModelBuilder, addNumericFeatureRef) {
   B.addConstraint(std::make_unique<ExcludesConstraint>(
-      std::make_unique<PrimaryFeatureConstraint>(AA),
-      std::make_unique<PrimaryFeatureConstraint>(AB)));
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ab"))));
```
Maybe a `using namespace std;` would make sense at the top of this file :wink:

vulder commented on unittests/Feature/FeatureModelBuilder.cpp:
```diff
@@ -157,47 +160,102 @@ TEST(FeatureModelBuilder, detectXMLAlternativesBroken) {
+          std::make_unique<BinaryFeature>("ab")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ac"))));
+  B.addConstraint(std::make_unique<ExcludesConstraint>(
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("aa")),
+      std::make_unique<PrimaryFeatureConstraint>(
+          std::make_unique<BinaryFeature>("ac"))));
```
This pattern looks repetitive, I think we should create a helper/short cut for this.

s9latimm replied:
I added a macro. 
With this the count of `std::` is significantly reduced, so I think we do not need to add `using namespace std` here.

vulder replied:
Please don't use macros :cry:, it should be possible to just build is as a constexpr function or maybe a template func.

vulder reviewed (changes requested):
What a big rewrite :laughing: 
But I guess the new design works?

padupr commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -443,30 +324,45 @@ template <typename... Rules> class FeatureModelConsistencyChecker {
+         }))) {
+      return true;
+    }
+    llvm::errs() << "Failed to validate 'CheckFeatureParentChildRelationShip'."
```
```suggestion
    llvm::errs() << "Failed to validate 'ExactlyOneRootNode'."
```

s9latimm committed:
Suggested changes

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp

s9latimm committed:
More suggested changes

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelBuilder.cpp

s9latimm committed:
Add DFS iterator

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Feature/OrderedFeatureVector.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	lib/Feature/OrderedFeatureVector.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/OrderedFeatureVector.cpp

s9latimm committed:
Suggested changes

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModelBuilder.cpp
M	lib/Feature/FeatureModelTransaction.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/NumericFeature.cpp
M	unittests/Feature/OrderedFeatureVector.cpp
M	unittests/Feature/Relationship.cpp
M	unittests/Feature/RootFeature.cpp
M	unittests/Feature/UnittestHelper.h
M	unittests/resources/test.xml

Commit:
Implement removal of features, relationships and locations

Commit:
Merge remote-tracking branch 'origin/f-FMBuilderMods' into f-FMBuilderMods

Commit:
Implement removal of features, relationships and locations

Commit:
TIL about IntelliJ's changelists

Commit:
add related issue to todos

vulder commented on lib/Feature/FeatureModel.cpp:
```diff
@@ -20,26 +25,25 @@ Feature *FeatureModel::addFeature(std::unique_ptr<Feature> NewFeature) {
-  FMB.setRootName(this->getRoot()->getName().str());
 
   for (const auto &KV : this->Features) {
     // TODO(s9latimm): Add unittests for cloned FeatureSourceRanges
```
TODO

s9latimm committed:
Address open TODOs

M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelTransaction.cpp
M	unittests/Feature/FeatureSourceRange.cpp
M	unittests/Feature/NumericFeature.cpp

s9latimm committed:
Replace assert with expect

M	unittests/Feature/FeatureSourceRange.cpp

s9latimm commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -63,13 +63,14 @@ class FeatureModelTransaction
-        // TODO(s9latimm): Committing now may break with prior failed commits.
-        //  We need a better way of tracking if a previous commit failed,
-        //  instead of checking locally if FM is nullptr.
+        // TODO(se-passau/VaRA/issues/744): Committing now may break with prior
```
```suggestion
        // TODO(se-passau/VaRA#744): Committing now may break with prior
```

s9latimm committed:
Update include/vara/Feature/FeatureModelTransaction.h

vulder committed:
Merge branch 'vara-dev' into f-FMBuilderMods
